### PR TITLE
[4.x] Add test for 103 handling

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

File: mockwebserver/src/main/kotlin/okhttp3/mockwebserver/MockResponse.kt [hunk 1]
```diff
@@ -347,6 +349,10 @@ class MockResponse : Cloneable {
 
   override fun toString() = status
 
+  fun addInformationalResponseInternal(informationalResponse: MockResponse): MockResponse = apply {
```
Comment (yschimke) [inline mockwebserver/src/main/kotlin/okhttp3/mockwebserver/MockResponse.kt:352]: This is not great, but seems better to have some limited support without the bigger change of backporting.

cc @swankjesse

File: okhttp/src/test/java/okhttp3/CallTest.java [hunk 2]
```diff
@@ -2812,6 +2812,43 @@ public void cancelWhileRequestHeadersAreSent_HTTP_2() throws Exception {
...
+                    .addInformationalResponseInternal(
+                            new MockResponse().setResponseCode(103).addHeader("Link", "</style.css>; rel=preload; as=style")
+                    )
+    );
+    Request request = new Request.Builder()
+            .url(server.url("/"))
+            .post(RequestBody.create("abc", MediaType.get("text/plain")))
+            .build();
+    executeSynchronously(request)
+            .assertFailure(SocketTimeoutException.class);
```
Comment (yschimke) [inline okhttp/src/test/java/okhttp3/CallTest.java:2829]: Showing we don't handle this well.

cc @swankjesse

**Review by swankjesse — APPROVED** [review]

File: mockwebserver/src/main/kotlin/okhttp3/mockwebserver/MockWebServer.kt [hunk 3]
```diff
@@ -1026,6 +1026,13 @@ class MockWebServer : ExternalResource(), Closeable {
         readBody = true
       }
 
+      peek.informationalResponses.forEach {
```
Comment (swankjesse) [inline mockwebserver/src/main/kotlin/okhttp3/mockwebserver/MockWebServer.kt:1029]: FWIW, I dislike `forEach` as an alternative to `for(...)`. It’s needlessly complex to step-debug!

File: okhttp/src/test/java/okhttp3/CallTest.java (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/test/java/okhttp3/CallTest.java:2829]: 👍🏻

**Review by swankjesse — APPROVED** [review]

**Pull request merged**